Support setting signal masks for child processes on Unix

sunshowers opened the pull request:
This commit implements basic support for setting signal masks during process spawning.
    
* With the `posix_spawn` code path, this means setting `posix_spawnattr_setsigmask`.
* With the `fork/exec` path, this means calling `pthread_sigmask` in the child before executing it.

This PR retains the default behavior (empty signal mask). I've also put up #101077 as a dependent PR to change the default behavior.

API change proposal:

* https://github.com/rust-lang/libs-team/issues/88

Unresolved questions:

* [ ] Is the possible UB with masking `SIGSEGV` etc (documented in the POSIX spec) enough reason to mark this unsafe?

sunshowers committed:
Implement sigdelset and sigismember for Android

We're going to need this for upcoming signal mask support.

sunshowers committed:
Support blocking and unblocking signals for child processes on Unix

This commit implements basic support for setting signal masks during process spawning.

* With the `posix_spawn` code path, this means setting `posix_spawnattr_setsigmask`.
* With the `fork/exec` path, this means calling `pthread_sigmask` in the child before executing it.

This also fixes a bug: previously, we were always setting the signal mask to
the empty set, breaking tools like `nohup`. With this change, we inherit
the signal mask from the parent by default.

rust-highfive commented:
r? @joshtriplett

(rust-highfive has picked a reviewer for you, use r? to override)

sunshowers commented:
Ah I'm not sure if there should be a reviewer at the moment, this is meant to be a draft PR :)

jyn514 commented:
@sunshowers I've unassigned a reviewer for now, but please do run `r? rust-lang/libs-api` once this is ready for review so it doesn't get lost :)

jyn514 commented on library/std/src/sys/unix/process/process_common.rs:
```diff
@@ -128,6 +169,47 @@ pub enum Stdio {
     Fd(FileDesc),
 }
 
+pub struct SignalSet {
```
Shouldn't this be private?

sunshowers replied:
It follows the pattern of the other items in the module which are all declared `pub` (though they actually resolve to `pub(crate)`).

joshtriplett commented on library/std/src/sys/unix/process/process_common.rs:
```diff
@@ -209,6 +293,9 @@ impl Command {
     pub fn pgroup(&mut self, pgroup: pid_t) {
         self.pgroup = Some(pgroup);
     }
+    pub fn signal_mask(&mut self, mask: SignalSet) {
```
This should accept an `Into<Option<SignalSet>>`, to allow passing `None` as well.

joshtriplett commented on library/std/src/sys/unix/process/process_common.rs:
```diff
@@ -209,6 +293,9 @@ impl Command {
     pub fn pgroup(&mut self, pgroup: pid_t) {
         self.pgroup = Some(pgroup);
     }
+    pub fn signal_mask(&mut self, mask: SignalSet) {
+        self.signal_mask = Some(mask);
```
```suggestion
    pub fn signal_mask(&mut self, mask: impl Into<Option<SignalSet>>) {
        self.signal_mask = mask.into();
```

sunshowers replied:
Thanks, done!

sunshowers commented:
Updates:
* Addressed review feedback from @joshtriplett
* Added documentation and examples

I think this is ready for review now.

sunshowers commented:
r? rust-lang/libs-api

sunshowers commented:
(I still need to write up the change proposal I guess!)

thomcc commented:
> Is the possible UB with masking SIGSEGV etc (documented in the POSIX spec) enough reason to mark this unsafe?

I guess the problem would be that it changes the documented contract of the `pre_exec` hook. I don't think there's a way to raise `SIGBUS`, `SIGFPE`, or `SIGSEGV` from safe code (well, a way that would violate the UB), but `SIGILL` is also listed as one of the signals that is UB to trigger when it's blocked, and we commonly use that for an "abort immediately" function (e.g. `core::intrinsics::abort()`). (Often this is used for "we need to abort and are inside libcore or liballoc rather than libstd")

That said, `pre_exec` is already unsafe and you have to be careful anyway, so I'm not 100% sure the extent to which this matters. The easy way of triggering sigill from safe code is a double panic, but inside `pre_exec`, the first panic will lead to `sys::abort_internal()`, e.g. `process::abort()` (and not `intrinsics::abort()`), which shouldn't raise SIGILL. (I guess a small caveat here is the `panic_immediate_abort` feature, which does use `intrinsics::abort()`).

I'm not sure if this is worth marking as unsafe, it seems unfortunate if so, but it's probably worth considering carefully if we aren't going to, to avoid a footgun. That said, there's probably enough time to get this right prior to stabilizing, so I don't know that this is a blocker.

sunshowers commented:
> That said, pre_exec is already unsafe and you have to be careful anyway, so I'm not 100% sure the extent to which this matters.

Yeah, that makes sense. Note that existing code won't be affected though -- only users of both `signal_mask` and `pre_exec` are potentially affected.

thomcc commented:
Yeah, so I suppose we could adjust the safety contract of `pre_exec` at the same time.

sunshowers commented:
The other case

sunshowers commented:
(whoops, closed by accident)

The other issue with safety I was thinking of is: what happens if the child itself has a segfault unrelated to the parent, and the parent has masked `SIGSEGV`? This isn't UB within Rust, but it is UB across the system. I don't know if unsafe extends to that kind of UB.

thomcc commented:
> I don't know if unsafe extends to that kind of UB.

In general it doesn't. It's totally safe to call processes that crash or do other UB. We have no way of preventing it, it's kind of outside Rust's ability to defend against in general.

That said, fork/exec stuff is ofc where the line gets blurry here (for example, technically `pre_exec` is run inside the child process, but it's still unsafe).

thomcc commented on library/std/src/sys/unix/process/process_common.rs:
```diff
@@ -43,17 +44,56 @@ cfg_if::cfg_if! {
+            }
+
+            let raw = slice::from_raw_parts_mut(set as *mut u8, size);
             raw[bit / 8] |= 1 << (bit % 8);
```
Hm, this isn't your fault, but these impls don't match the cited [android impls](https://github.com/aosp-mirror/platform_bionic/blob/ad8dcd6023294b646e5a8288c0ed431b0845da49/libc/include/android/legacy_signal_inlines.h). In particular, it uses `c_long` rather than u8. (`LONG_BIT` is `size_of::<c_long>() * 8`, but we just are using `8`)

I think this will result in incorrect behavior on big endian android, although it's not hard to guess why nobody's noticed before.

sunshowers replied:
Thanks for catching this! Once #100782 lands I'll rebase and port the changes there over here. (Also I noticed you fixed up the error handling in that PR; thanks for doing that!)

zackw commented:
What I'm about to describe is a pre-existing bug, but it would be nice either to fix it along with this PR, or else, at least, not make it worse:

It is _incorrect_ to set the signal mask to the empty set in child processes.  This breaks `nohup` and other tools that intentionally spawn a process with signals blocked, expecting that setting to inherit to all processes transitively spawned by that one.  The correct behavior is to _pass down_ the parent process's signal mask, whatever it is, _modified_ by any preferences expressed by the parent.

The `posix_spawnattr_setsigmask` API is expected to be used like this:

```c
sigset_t ss;
pthread_sigmask(SIG_BLOCK, NULL, &ss); // read current thread's signal mask into `ss`
sigaddset(&ss, SIGINT); // for instance
posix_spawnattr_setsigmask(&attr, &ss);
```

Therefore I'd like to ask you to make the following changes:

- Fix the existing bug.  When your new API is not used, the child process's signal mask should default to whatever the parent process's signal mask is, _not_ to empty.  (I was going to say that it should be the parent process's signal mask after undoing any changes made by the Rust runtime on startup, but it looks like the Rust runtime no longer messes with the signal mask, so we're good on that front already.)
- Define your new API in terms of _making changes relative to the parent process's signal mask_, rather than in terms of "here is the brand new signal mask to use."  That could, for instance, look something like this:

    ```rust
    #[unstable(feature = "process_sigmask", issue = "none")]
    fn block_signal(&mut self, sig: std::os::unix::signal::Signal) -> &mut process::Command;
    #[unstable(feature = "process_sigmask", issue = "none")]
    fn unblock_signal(&mut self, sig: std::os::unix::signal::Signal) -> &mut process::Command;
    ```

    where `std::os::unix::signal::Signal` is an enum of all the valid signal constants.

sunshowers commented:
Thanks for the review, @zackw. I think that makes sense. Note that I will be away for this week and busy next week, so I'll only be able to look at this starting the week of 2022-09-05.

Can the signal to block/unblock just be an i32? The list of signals is pretty long and varies by system, and it seems like letting people use signals from the libc crate would be easiest.

zackw commented:
> Can the signal to block/unblock just be an i32? The list of signals is pretty long and varies by system, and it seems like having libc be the source of truth and letting people use signals from there would be easiest.

The only problem with that is you have to decide what to do if you are given an argument that's not a valid signal number (`libc::sigaddset` fails).  Because this is a builder, it's not convenient to return a Result.  I _think_ it would be okay to ignore the error, because the most plausible reason for this to happen is if someone's trying to mask a signal that exists on only some systems.  If the signal doesn't exist, then it can't happen, so it doesn't matter if it's masked or not.  But I'd like to know what other people think here.

sunshowers commented:
> Because this is a builder, it's not convenient to return a Result.

Hmm, is this true? I'm not sure what's generally done in std, but I've definitely written code where consumers have to do:

```rust
MyBuilder::new()
    .fallible_method(arg)
    .expect("arg is valid")
    .fallible_method_2(arg2)?
    .infallible_method()
    .build()
```

joshtriplett commented:
I agree, the method should return an error.

bors commented:
:umbrella: The latest upstream changes (presumably #100782) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

sunshowers commented:
Did the following:

* Rebased and incorporated the changes from #100782
* Implemented a new API based on the suggestion from @zackw with some changes (in particular, I added a `will_block_signal` that I think is really useful).
* Updated the API change proposal with this new API
* Split the diff into two commits: the first one to add the new methods to the Android impl, and the second commit to implement signal mask support.

sunshowers commented:
r? rust-lang/libs-api

sunshowers commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

sunshowers commented:
@rustbot label -S-experimental

thomcc commented:
There is no libs-api -- it uses the teams from this list (and ignores any prefix like `rust-lang/`): https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/rust.json#L2-L95. The API review for this will happen on the ACP tho, so I don't mind reviewing the impl, since I think I might have the most review bandwidth at the moment.

r? @thomcc

thomcc commented on library/std/src/sys/unix/process/process_unix.rs:
```diff
@@ -326,18 +326,10 @@ impl Command {
-            cvt(sigemptyset(set.as_mut_ptr()))?;
-            cvt_nz(libc::pthread_sigmask(libc::SIG_SETMASK, set.as_ptr(), ptr::null_mut()))?;
+            // Set the signal mask.
+            let signal_mask = self.get_signal_mask()?;
```
This is a behavioral change we should consider separately. If we're going to make it here, then it needs an FCP too.

sunshowers replied:
Moved to #101077.

thomcc commented on library/std/src/sys/unix/process/process_unix.rs:
```diff
@@ -529,11 +521,12 @@ impl Command {
-            cvt_nz(libc::posix_spawnattr_setsigmask(attrs.0.as_mut_ptr(), set.as_ptr()))?;
-            cvt(sigaddset(set.as_mut_ptr(), libc::SIGPIPE))?;
-            cvt_nz(libc::posix_spawnattr_setsigdefault(attrs.0.as_mut_ptr(), set.as_ptr()))?;
+            let signal_mask = self.get_signal_mask()?;
```
ditto.

sunshowers replied:
Moved to #101077.

thomcc reviewed (changes requested):
This contains behavioral changes -- the ones suggested in https://github.com/rust-lang/rust/pull/100737#issuecomment-1222672934. So it needs an FCP (which would be a bit confusing since it's an unstable API), or for that to be split into it's own thing.

It's your call, I think the switch to sigemptyset to reading the current signal mask should be split into its own PR (which should be doable, unless I've missed something).

sunshowers commented:
OK. Well, I was just going by what @zackw said above. I'll back that change out. However, I would like someone else (maybe @zackw, since they suggested the change) to initiate the FCP.

sunshowers commented:
I've put up #101077 as a dependent PR on top of this one.

thomcc commented on library/std/src/sys/unix/process/process_common.rs:
```diff
@@ -296,6 +396,10 @@ impl Command {
     }
+    #[allow(dead_code)]
+    pub fn get_signal_mask(&self) -> io::Result<&SignalSet> {
+        self.signal_mask.get_or_try_init(SignalSet::empty)
```
We'll want to remove the OnceLock from this if the other PR's FCP doesn't get accepted.

thomcc reviewed (commented):
The impl looks fine although I'll probably want to review again after the situation with the other PR is resolved.

thomcc commented:
I'm going to mark this as blocked on the other PR, as it's unlikely that we want this impl without that (and the resolution there may change how we approach this API).

bors commented:
:umbrella: The latest upstream changes (presumably #100786) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jsgf commented on library/std/src/os/unix/process.rs:
```diff
@@ -179,6 +179,107 @@ pub trait CommandExt: Sealed {
+    /// success and has no effect.
+    ///
+    /// Blocking some signals like `SIGSEGV` can lead to undefined behavior in
+    /// the child. See the [`pthread_sigmask`] man page for more information.
```
sigprocmask is probably the more canonical reference.

Also, while the manpage may say that the effect of blocking synchronous traps (ie, SEGV/BUS/FPE/ILL raised while executing an instruction) is undefined, I wouldn't read that as being the same as UB at the Rust/llvm level.

In practice I think it would either deliver the signal as SIG_DFL (ie, Linux), or endlessly re-trap on the same instruction (what x86 macOS seems to do) - either way the program has effectively terminated and won't enter any UB states as far as the program semantics are concerned.

sunshowers replied:
Heh it's actually the same man page for both `pthread_sigmask` and `sigprocmask`. I can change the reference in the docs to be `sigprocmask` though.

jsgf replied:
Huh - on Fedora they're separate, but pthread_sigmask has almost no content and refers to sigprocmask for discussion of the detailed semantics.

sunshowers replied:
(I was referring to the man pages from the official POSIX spec here, as linked in the comment)

thomcc commented:
I think this is no longer blocked.

sunshowers commented:
Yep, planning to get to it soon. Meanwhile I've worked around it on beta Rust by calling pthread_sigmask before and after process spawning, which appears to work.

JohnCSimon commented:
@sunshowers 
ping from triage - can you post your status on this PR? This has sat unmoved for a while now. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

Dylan-DPC commented:
Closing this due to inactivity. Feel free to make a new pr if you wish to continue this.

sunshowers commented:
Apologies, I started a new full-time job since I wrote this and haven't had the time to follow up on it. After #101077, I also figured out a way to do this without needing to add new API surface to std:

1. Temporarily use `pthread_sigmask` to set a new sigmask.
2. Spawn the process.
3. Unset the sigmask.

(and use a scope guard to ensure that the sigmask is unset in case of panic.)

This means that I'm probably not going to pursue this further. However, someone else is welcome to.
